### ISSUE TITLE
Wait… When Grow() is called, then _buffer is now pointing to an array that was just returned to the pool?

### ISSUE THREAD

**Issue by toupswork** [linkdotnet/BlogExamples#3]
Just discovered you, and I am already a huge fan. Thank you for these wonderful articles.

Did you mean to have the _buffer span pointing to a buffer that was just returned to the pool? Doesn't that mean if it's recycled elsewhere it can be pointing to different values? Or is there something I am not aware of?

**Comment by linkdotnet** [comment]
Hey @toupswork  - thank you for the kind words.

And on second look - yes that code snippet is wrong. I did a full blown version of that `ValueStringBuilder` which indeed fixes that behavior, see: https://github.com/linkdotnet/StringBuilder/blob/5ad3e8a0159d0cd87ff18cb5bb09de7fc4ab5cd7/src/LinkDotNet.StringBuilder/ValueStringBuilder.cs#L306

The example "works" as there is no competition around the array pools. I will correct that bid tonight.

Thank you very much again!

**Comment by linkdotnet** [comment]
Tonight, as in 3 days later, but I managed to fix the code samples as well as the blog post itself.
Thanks again @toupswork !

**Comment by toupswork** [comment]
Nice! Seems like you got it covered. I can't find anything wrong with it. Probably the only thing I would do different is initialize that first 32 chars from the stack rather than the heap. I know you can't `stackalloc` in the ctor, but you could probably use the C# 12 inline array to get that from the stack.  Then maybe get rid of the dispose, and whenever `ToString()` is called the first time, store the reference to the allocated string and then return the last rented buffer to the pool. Then if `ToString()` is called again, it would return the stored instance.  But if someone were to try to append to it after calling `ToString()`, I'd rent a buffer from the pool again and start over.  Not asking you to do all that, just enjoy sharing ideas.

**Comment by linkdotnet** [comment]
That is a good point. 
You can achieve somewhat the same, when asing for a`Span<char>` from a consumer.
See: https://github.com/linkdotnet/StringBuilder/blob/5ad3e8a0159d0cd87ff18cb5bb09de7fc4ab5cd7/src/LinkDotNet.StringBuilder/ValueStringBuilder.cs#L39

Unfortunately, I can not move that discussion towards the other repository. Anyway - the InlineArray would at least somehow enforce stack allocation.

In the library, I pushed that responsibility toward the user. 
I also wanted to experiment with a rope-like structure - but until now, C# doesn't allow references to `ref struct`s inside a ref struct making it impossible to experiment with this.

If you have any further ideas to bounce back and forth,I am very much interested.

**Comment by toupswork** [comment]
That would be very cool. The thing that comes to my mind with that idea is that because strings can get long, you'd risk eating up a lot of stack space, even if the struct is just holding pointers to a string on the heap. I can't help but wonder what the technical reason is for you not wanting to just use a regular struct. I would assume you're worried it could get copied to the heap by boxing or someone making it a field on a class. Or maybe getting it captured in a closure.

**Comment by linkdotnet** [comment]
> you'd risk eating up a lot of stack space
Yes that is true. My way of thinking is that in this case you are "not a normal user" of an API, but you are down the rabbit hole of optimizing where you know what you are doing.
That said, if you don't know the rough size of the generated string, you can set the initial size rather small. When the `Grow` method is called the next time, a buffer from the `ArrayPool` is rented, minimizing the risk of allocating too much stack space.

Basically - yes, `ref struct` is just enforcement for the stack and avoiding scenarios where you can run into weird situations  as you described (for example, boxing, async/await, closures, and so on).

The general tone of the library is not to be a replacement for `System.Text.StringBuilder`.
